### ISSUE TITLE
Dithering 

### ISSUE THREAD

**Issue by DarkAlchy** [tannerhelland/PhotoDemon#258]
The nightly build dithering has an issue and is now lost a lot of quality.

I am not sure if it is a loss of quality but something sure has changed and for engraving work it is now just not usable.

![old](https://user-images.githubusercontent.com/10561690/39067499-62e3bff8-44c8-11e8-950e-07fcb5795482.jpg)
That is the old dither which was perfect for real world work like laser engraving but here comes the new one and YIKES!  It is more stylized (look at the sky for instance) but engraves horribly.
![new](https://user-images.githubusercontent.com/10561690/39067540-891f791e-44c8-11e8-82ba-05637fa59ba6.jpg)
Everything is set exactly the same and both are using Stucki (all of the algorithms changed like this).

**Comment by tannerhelland** [comment]
Thanks for starting this issue, @DarkAlchy.  A few quick comments...

1) What tool are you using to apply the dither?  There are three different tools that provide this feature (Adjustments > Black and White, Adjustments > Monochrome > Color to monochrome, and Effects > Stylize > Palettize).  

2) On all three of those tools, if you bump the "dithering strength" up to its maximum value, you should get identical output to the old tool.  (The biggest difference now is that you can *reduce* dithering strength to get better results on images with sharp contrast.)  So if you still want the old results, just bump dithering strength up to 100%.

3) When you're very zoomed-out from a dithered image (like in the screenshots you shared), PhotoDemon has to "downsample" the image to fit it on-screen.  This can cause the dithering to look blurry and weird.

If you set zoom to 100%, you'll get a better idea of the actual dithering results, including what they will look like in an engraving.  (Engravers obviously have much higher DPI support than the average computer display does, so there's no perfect way to preview them on a screen.  100% zoom is the closest thing we've got.)

**Comment by DarkAlchy** [comment]
Yes, these things I know but here is the sky, for instance, at 100% and you can see the error not being distributed correctly or whatever those streaks are.

Old
![image](https://user-images.githubusercontent.com/10561690/39068440-d31f17d8-44cb-11e8-9083-4e6e8653d119.png)
New
![image](https://user-images.githubusercontent.com/10561690/39068457-df6b2cca-44cb-11e8-9558-4612899c70ec.png)

The image is also way overly dark now and how I got here, on both, was via Adjustments->Black and white and as you can see they are not the same (of course I figured 100% on the slider was supposed to be the same as the old way but I did play with it).

**Comment by tannerhelland** [comment]
Thank you for the follow-up.  Can you attach the elephant image you are using?  I will take a closer look to see if I can reproduce your results.

**Comment by DarkAlchy** [comment]
Sure can.
![mammal-3218712](https://user-images.githubusercontent.com/10561690/39068812-19e0b874-44cd-11e8-8dfb-0b8523ff4d35.jpg)

**Comment by DarkAlchy** [comment]
Oh, I just tried Adjustments > Monochrome > Color to monochrome and that one works so they are identical at 100%.

**Comment by DarkAlchy** [comment]
I do have another oddity and this is Atkinson at 100% for both (old then new)

![image](https://user-images.githubusercontent.com/10561690/39070155-bfa0fb8a-44d1-11e8-8794-04dc567b9ae0.png)
![image](https://user-images.githubusercontent.com/10561690/39070162-c4382290-44d1-11e8-9ddc-fe6f40e83767.png)
See those patterns and the weird stuff by the birds?  It has shifted to the left which may have caused the other patterns to change so I am only concentrating on the stuff by the birds.

**Comment by tannerhelland** [comment]
Okay, thank you for the extra information.  I've made some changes to the code and uploaded a new developer build.  (I'll spare any further explanation until we see if this improves quality!)

Download link again: http://photodemon.org/downloads/nightly/PhotoDemon_nightly.zip

**Comment by DarkAlchy** [comment]
Atkinson is still doing the above so that wasn't touch yet so let me test the other ways to dither.

**Comment by DarkAlchy** [comment]
It appears that the other ways to dither now work as well.

**Comment by DarkAlchy** [comment]
I will say that Atkinson is not the same across the versions and if I did an overlay it would show the differences.

**Comment by tannerhelland** [comment]
Awesome, thank you for testing this latest version again!  Sorry that it's taken me a few tries to get everything right.  😥 

To me, the Atkinson changes are okay and expected.  If you check some other demonstrations (e.g. https://gazs.github.io/canvas-atkinson-dither/ ), they produce a very similar pattern by the birds, with only slight variations that you'd expect given different implementation details.  (Atkinson dithering was originally used on old Macintosh computers, and a "diagonal pattern" is an intentional part of its design.)  I hope that's okay.

With the latest changes, the dithering algorithm *is* slightly changed from the old one.  The old version didn't calculate diffusion as accurately as it could.  The new one should be better in that regard, although it will lead to subtle differences.  Specifically, I've tweaked the greyscale tool to work a little differently from the monochrome and color ones, although for pure black-and-white dithering, I imagine the monochrome tool is still capable of the best results, because you can adjust the threshold to control the "split point" between black and white.  

If you see any other glaring problems, please let me know and I'll revisit again.

Also - I really appreciate your help and feedback on this feature.  I'd be happy to add your name to the contributor list for the next release in exchange for everything you've done.  Just let me know what name (and/or URL) you'd like me to use, and I'll get it added.  Thank you again!

**Comment by DarkAlchy** [comment]
You are welcome.  I have no URL  and just use the name shown. 

Thank you for this program.

A question but when I use the magic wand tool I see no way to add, or subtract, another selection.  I found this when I was taking the background out of a picture and I couldn't just add a selection (the space between their legs) to my current selection no matter what I tried.  I am used to PS and that is a really simple task to do.
